core/trino-main/src/test/java/io/trino/sql/planner/optimizations/TestPlanNodeSearcher.java

```diff
@@ -0,0 +1,97 @@
+ * limitations under the License.
+ */
+package io.trino.sql.planner.optimizations;
+
+import com.google.common.collect.ImmutableList;
+import io.trino.metadata.AbstractMockMetadata;
+import io.trino.sql.planner.PlanNodeIdAllocator;
+import io.trino.sql.planner.iterative.rule.test.PlanBuilder;
+import io.trino.sql.planner.plan.Assignments;
+import io.trino.sql.planner.plan.JoinNode;
+import io.trino.sql.planner.plan.PlanNode;
+import io.trino.sql.planner.plan.PlanNodeId;
+import io.trino.sql.planner.plan.ProjectNode;
+import org.testng.annotations.Test;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.sql.planner.plan.JoinNode.Type.INNER;
+import static io.trino.testing.assertions.Assert.assertEquals;
+
+public class TestPlanNodeSearcher
+{
+    private static final PlanBuilder BUILDER = new PlanBuilder(new PlanNodeIdAllocator(), new AbstractMockMetadata() {});
+
+    @Test
+    public void testFindAll()
+    {
+        int size = 10;
+        ProjectNode root = BUILDER.project(Assignments.of(), BUILDER.values());
+        for (int i = 1; i < size; i++) {
+            root = BUILDER.project(Assignments.of(), root);
+        }
+
+        List<PlanNodeId> rootToBottomIds = new ArrayList<>();
+        PlanNode node = root;
+        while (node instanceof ProjectNode) {
+            rootToBottomIds.add(node.getId());
+            node = ((ProjectNode) node).getSource();
+        }
+
+        List<PlanNodeId> findAllResult = PlanNodeSearcher.searchFrom(root)
+                .where(ProjectNode.class::isInstance)
+                .findAll()
+                .stream()
+                .map(PlanNode::getId)
+                .collect(toImmutableList());
+
+        assertEquals(rootToBottomIds, findAllResult);
+    }
+
+    @Test
+    public void testFindAllMultipleSources()
+    {
+        List<JoinNode> joins = new ArrayList<>();
+        for (int i = 0; i < 4; i++) {
+            joins.add(BUILDER.join(INNER, BUILDER.values(), BUILDER.values()));
+        }
+        JoinNode leftSource = BUILDER.join(INNER, joins.get(0), joins.get(1));
+        JoinNode rightSource = BUILDER.join(INNER, joins.get(2), joins.get(3));
+        JoinNode root = BUILDER.join(INNER, leftSource, rightSource);
+
+        ImmutableList.Builder<PlanNodeId> idsInPreOrder = ImmutableList.builder();
+        preorder(root, idsInPreOrder);
+
+        List<PlanNodeId> findAllResult = PlanNodeSearcher.searchFrom(root)
+                .where(JoinNode.class::isInstance)
+                .findAll()
+                .stream()
+                .map(PlanNode::getId)
+                .collect(toImmutableList());
+
+        assertEquals(idsInPreOrder.build(), findAllResult);
+    }
+
+    private static void preorder(PlanNode root, ImmutableList.Builder<PlanNodeId> builder)
+    {
+        if (root instanceof JoinNode) {
```

Comment:
Yes, this method is supposed to collect the root id and immediately contained JoinNodes' ids in pre-order. It is just for the purpose of this single test, and it is simplified to match the actual example.
However, if it seems broken, I'll support visiting the whole tree and rename the method to a more specific name.

Reply:
No need to make it visit the whole tree. But, at a minimum, make it fail if a node is not a JoinNode. Otherwise, it will be surprising to future callers if the method silently ignores input but doesn't produce the "expected" result.